core/indigo-core/reaction/src/reaction_multistep_detector.cpp

```diff
@@ -238,6 +227,430 @@ void ReactionMultistepDetector::createSummBlocks()
+                    continue;
+                auto zone = isMergeable(qel.first, j, qel.second);
+                if (zone.has_value())
+                {
+                    cluster.push_back(j);
+                    bfs_queue.emplace(j, zone);
+                }
+            }
+        }
+        for (std::size_t k = 1; k < cluster.size(); ++k)
+        {
+            QS_DEF(Array<int>, mapping);
+            if (_components[cluster[k]].first)
+            {
+                _components[i].first->mergeWithMolecule(*_components[cluster[k]].first, &mapping, 0);
+                _components[cluster[k]].first.reset();
+            }
+        }
+    }
+    _components.erase(std::remove_if(_components.begin(), _components.end(), [](auto& p) { return !p.first; }), _components.end());
+    _moleculeCount = (int)_components.size();
+}
+
+std::optional<std::pair<int, int>> ReactionMultistepDetector::isMergeable(size_t mol_idx1, size_t mol_idx2, std::optional<std::pair<int, int>> current_zone)
+{
+    auto& mdi = _mol_distances[mol_idx1];
+    auto dist_it = mdi.distances_map.find(mol_idx2);
+    if (dist_it != mdi.distances_map.end() && dist_it->second < LayoutOptions::DEFAULT_BOND_LENGTH * 2)
+    {
+        // collect surrounding zones for both molecules
+        std::map<int, std::unordered_set<int>> other_zones1;
+        std::map<int, std::unordered_set<int>> other_zones2;
+        std::map<int, std::unordered_set<int>> comm_zones;
+
+        auto zone1 = findMaxSpecialZone(mol_idx1, other_zones1);
+        auto zone2 = findMaxSpecialZone(mol_idx2, other_zones2);
+        std::set_intersection(other_zones1.begin(), other_zones1.end(), other_zones2.begin(), other_zones2.end(), std::inserter(comm_zones, comm_zones.begin()),
+                              [](auto& a, auto& b) { return a.first < b.first; });
+
+        // no zone - no reason to merge
+        if (zone1.has_value())
+        {
+            // if both molecules has the same zone and section - merge
+            // we never merge molecules with different sections of the same zone
+            if (zone2.has_value())
+            {
+                // disable catalysts merging
+                if ((_zones[zone1.value().first].zone_type == ZoneType::EArrow && zone1.value().second > 1) ||
+                    (_zones[zone2.value().first].zone_type == ZoneType::EArrow && zone2.value().second > 1))
+                    return std::nullopt;
+
+                // if both molecules are on the same zone - check if they are mergeable
+                for (auto& cz : comm_zones)
+                {
+                    auto it_oz2 = other_zones2.find(cz.first);
+                    if (it_oz2 != other_zones2.end())
+                    {
+                        for (auto& section : cz.second)
+                        {
+                            if (!(_zones[cz.first].zone_type == ZoneType::EPathWay && section > 1) && it_oz2->second.count(section ^ 1))
```

Comment:
Could you please add comment to this check?
For me not clear what is checked here.